go-client/client/tablehandle.go

```diff
@@ -51,6 +60,49 @@ func newTableHandle(client *Client, ticket *ticketpb2.Ticket, schema *arrow.Sche
 	}
 }
 
+// newTableHandle returns a new table handle given information about the table,
+// typically from a gRPC ExportedTableCreationResponse.
+// If WithTableLeaksAllowed was not provided when creating a client,
+// this will also attaches a finalizer to the TableHandle that will warn if the handle is leaked and free the table automatically.
+func newTableHandle(client *Client, ticket *ticketpb2.Ticket, schema *arrow.Schema, size int64, isStatic bool) *TableHandle {
+	handle := newBorrowedTableHandle(client, ticket, schema, size, isStatic)
+
+	if !client.allowLeakedTables {
+		stackTrace := string(debug.Stack())
+
+		runtime.SetFinalizer(handle, func(th *TableHandle) {
+			// Start up a new goroutine, since finalizers block the GC
+			// and this needs to lock a mutex and perform a network request.
+			go func() {
+				th.lock.Lock()
+				defer th.lock.Unlock()
+
+				if th.client != nil { // using IsValid here would cause deadlock
+					th.client.lock.Lock()
+					defer th.client.lock.Unlock()
```

Comment:
It feels kind of dirty to directly grab the client lock from a table handle method.